### PR TITLE
8301989: new javax.swing.text.DefaultCaret().setBlinkRate(N) results in NPE

### PR DESCRIPTION
**Pull request by azuev-java**
Check if the component is associated with the caret before calling methods from it. Added test case that will make sure that will not happen again.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301989](https://bugs.openjdk.org/browse/JDK-8301989): new javax.swing.text.DefaultCaret().setBlinkRate(N) results in NPE

### Reviewers
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Author)
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - Committer)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk20 pull/122/head:pull/122` \
`$ git checkout pull/122`

Update a local copy of the PR: \
`$ git checkout pull/122` \
`$ git pull https://git.openjdk.org/jdk20 pull/122/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 122`

View PR using the GUI difftool: \
`$ git pr show -t 122`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk20/pull/122.diff">https://git.openjdk.org/jdk20/pull/122.diff</a>

</details>

### CODE REVIEW

File: src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java [hunk 1]
```diff
@@ -1050,7 +1050,7 @@ public void setBlinkRate(int rate) {
             throw new IllegalArgumentException("Invalid blink rate: " + rate);
         }
         if (rate != 0) {
-            if (component.isEditable()) {
+            if (component != null && component.isEditable()) {
                 if (flasher == null) {
```
Comment (prrace) [inline src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: So later if there is a non-editable component, what happens ?
Some existing logic for that will kick in ?

Comment (azuev-java) [reply src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: Yes, when the the caret is to be assigned to a component it will reset the blink rate and the logic will kick in. Manually tested with the possible scenarios like adding custom set blink rate cursor to a non-editable component and then turn it into editable later, add to editable and switch to non-editable later and so on.

**Review by DamonGuy — APPROVED** [review]
The changes seem to fix the NPE as expected. When testing, the provided test fails before the patch and passes after the patch is applied.

I was curious if the 2nd conditional was necessary to check for "component == null", but it definitely is. Also, your test checks for this, so that helped clear it up.

**Review by honkar-jdk — APPROVED** [review]
The patch works as expected. NPE problem is resolved. Additionally checked the fix by adding DefaultCaret to component (JTextField) in enabled and disabled states, it works as expected - Blinking caret when enabled and no caret when disabled.

File: src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java (hunk 1 above)
Comment (prrace) [reply src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: So if I go back and look at the code before your previous fix, there was no check,
so the timer is created and then I'd expect applies when the caret is installed on a component.
Here that behaviour is changed, so that if we create a caret, then call setBlinkRate(), and then install()
then the blink rate appears to be ignored whereas before 4512626 it was applied.

Comment (azuev-java) [reply src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: > if we create a caret, then call setBlinkRate(), and then install()
> then the blink rate appears to be ignored whereas before 4512626 it was applied.

Ok, if the text component is already visible and editable and focused and we install the new caret then the blink rate will not have an effect until the focus is lost and regained or until it became non-editable - and i think that was close to what we had before and there is a bug to investigate and optimize such corner case. In all other cases the code works fine. The value set is being preserved and will be reactivated when any of the methods such as focusGained or when the component installUI is called - if the cursor is set before component becomes visible - will cause to re-evaluate the blink rate and the stored value set by user will take place.

Comment (prrace) [reply src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: Suppose I have this code

Caret c = new DefaultCaret()
c.setBlinkRate(50);
c.install(editableJComponent);

BEFORE 4512626 the JDK code looked like this

     if (rate != 0) {
            if (flasher == null) {
                flasher = new Timer(rate, handler);
            }
            flasher.setDelay(rate);
        }


With your fix it looks like this

     if (rate != 0) {
            if (component != null && component.isEditable()) {
               if (flasher == null) {
                   flasher = new Timer(rate, handler);
               }
             }
        }


So with my sample application code there's a change that my blink rate is completely lost.

Comment (prrace) [reply src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: So it seems that in such a case, "savedBlinkRate" is still set and that is used to both report the value of blink rate *and* create & start a Timer when it is needed.
It is a little hard to follow through all those "when it is needed" cases but if we are attached to an enabled & editable component when  focusGained() is called it should then start the flasher timer.

Comment (azuev-java) [reply src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:1054]: Using the simple code like this one:
```        Caret caret = new DefaultCaret();
        caret.setBlinkRate(10);
        JFrame frame = new JFrame("test");
        JTextArea area = new JTextArea(null, "Some text to make it non-empty", 5, 40);
        area.setEditable(false);
        area.setCaret(caret);
        frame.setLayout(new BorderLayout());
        frame.add(area, BorderLayout.CENTER);
        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        frame.pack();
        frame.setVisible(true);
        SwingUtilities.invokeAndWait(() -> area.requestFocus(false));
        SwingUtilities.invokeAndWait(() -> area.setEditable(true));
        SwingUtilities.invokeAndWait(() -> {
            System.out.println(area.getCaret().getBlinkRate());
        });
```
i tested many possible combinations to see that the blink rate specified in setBlinkRate is preserved and not being overwritten by some code. In some corner cases the cursor does not blink despite reporting the correct blink rate until the focus is lost and regained - but that is still better than the version before JDK-4512626 where in the same corner case the cursor is not visible until the focus is lost and gained again. There is a bug JDK-8299048 that tracks the corner case i am talking about.

**Comment by prrace** [comment]
Another question, in the branch where the app calls
setBlinkRate(0)
we have
            if ((component == null || component.isEditable()) && isBlinkRateSaved) {
                savedBlinkRate = 0;
                isBlinkRateSaved = false;
            }

Why do we not clear these unconditionally ?
What is the point of saving it if there's a non-editable component ?

**Comment by azuev-java** [comment]
> Why do we not clear these unconditionally ?
> What is the point of saving it if there's a non-editable component ?

Because then when we set component non-editable the stored blink rate will be lost and we will not be able to restore it when component will be switched to editable again. Same with the component not assigned - if we assign it to the non-editable component and then flip the editable property then we would not have the previous blink rate saved and will not know what value to set.

**Comment by prrace** [comment]
> > Why do we not clear these unconditionally ?
> > What is the point of saving it if there's a non-editable component ?
> 
> Because then when we set component non-editable the stored blink rate will be lost and we will not be able to restore it when component will be switched to editable again. Same with the component not assigned - if we assign it to the non-editable component and then flip the editable property then we would not have the previous blink rate saved and will not know what value to set.

I am not following .. this is about the branch where the app is setting blink rate to zero.
We surely don't want to keep the old blink rate just because a component is non-editable ?

**Comment by azuev-java** [comment]
> We surely don't want to keep the old blink rate just because a component is non-editable ?

When we receive focus on non-editable component we set blink rate to 0 to stop caret blinking. If we do not add this condition then we immediately clear off the saved blink rate.

**Comment by prrace** [comment]
> > We surely don't want to keep the old blink rate just because a component is non-editable ?
> 
> When we receive focus on non-editable component we set blink rate to 0 to stop caret blinking. If we do not add this condition then we immediately clear off the saved blink rate.

How do you tell that apart from the app directly clearing it ? ie the APP calling setBlinkRate(0) ?

**Comment by azuev-java** [comment]
> How do you tell that apart from the app directly clearing it ? ie the APP calling setBlinkRate(0) ?

I do not - while component is non-editable you can not change blink rate to 0, once the component will become editable the last non-zero value will be restored.

**Comment by prrace** [comment]
> > How do you tell that apart from the app directly clearing it ? ie the APP calling setBlinkRate(0) ?
> 
> I do not - while component is non-editable you can not change blink rate to 0, once the component will become editable the last non-zero value will be restored.

But that seems wrong. Why is it not wrong ?

**Comment by azuev-java** [comment]
> But that seems wrong. Why is it not wrong ?

It might be an oversight of the initial fix, if you think it is wrong we can submit a bug and i can fix it. I definitely do not think it is critical and i do not think it is related to the nature of this change.

**Comment by prrace** [comment]
> > But that seems wrong. Why is it not wrong ?
> 
> It might be an oversight of the initial fix, if you think it is wrong we can submit a bug and i can fix it. I definitely do not think it is critical and i do not think it is related to the nature of this change.

Well .. it does seem wrong to me .. I was hoping you'd be able to explain why its OK.
That makes two "corner cases" to clean up in a later release.
Dilemma .. way too risky to try to clean that up now .. not sure if we want to back out the ENTIRE fix ... not sure about OKing a fix with known issues.
The only way I can look at it and convince myself is to consider that that corner case itself does not look like a P2 release stopper ..

**Review by prrace — APPROVED** [review]
Approving as critical to fix the NPE.
There seems possible there may be followup work in a later release.

**Comment by azuev-java** [comment]
> Approving as critical to fix the NPE. There seems possible there may be followup work in a later release.

There will be.
